pkg/sql/backfill.go

```diff
@@ -1536,8 +1536,14 @@ func ValidateConstraint(
 			return ie.WithSyntheticDescriptors(
 				[]catalog.Descriptor{tableDesc},
 				func() error {
-					return validateCheckExpr(ctx, &semaCtx, txn, sessionData, ck.GetExpr(),
+					err := validateCheckExpr(ctx, &semaCtx, txn, sessionData, ck.GetExpr(),
 						tableDesc.(*tabledesc.Mutable), ie, indexIDForValidation)
+					// Re-code the error to NotNullViolation if it's a NOT NULL constraint violation.
+					if err != nil && pgerror.GetPGCode(err) == pgcode.CheckViolation &&
+						ck.IsNotNullColumnConstraint() {
+						err = pgerror.WithCandidateCode(errors.UnwrapOnce(err), pgcode.NotNullViolation)
```

Comment:
This `UnwrapOnce` is relying on the outermost error being the pgerror. That feels fragile. Maybe better to use `errors.If` to find the correct pgerror and replace it directly?